### PR TITLE
MM-57853: Whitelist Restricted (simplified)

### PR DESCRIPTION
**Pull request by calebroseland**
#### Summary
- connected users config simplified
- bots: no invites
- guests: no invites, no notify-connect client request
- whitelist: remove user from whitelist when invited
- if restricted, must be whitelisted to connect or receive invite
- plugin bot can always connect

image

#### Screenshots
`/msteams connect`
CleanShot 2024-04-29 at 15 56 58

(no visible msg for restricted invite flow, but verified no invite received)

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-57853

### CODE REVIEW

File: server/command.go [hunk 1]
```diff
@@ -496,26 +500,6 @@ func (p *Plugin) executeConnectBotCommand(args *model.CommandArgs) (*model.Comma
...
-		if openConnectErr != nil {
-			p.API.LogWarn("Error in checking if the bot user can openly connect", "bot_user_id", p.userID, "error", openConnectErr.Error())
-			return p.cmdError(args, genericErrorMessage)
-		}
-
-		if !canOpenlyConnect {
-			return p.cmdError(args, "You cannot connect the bot account because the maximum limit of users allowed to connect has been reached.")
-		}
-	}
-
```
Comment (calebroseland) [inline server/command.go:518]: The built-in plugin bot should always be permitted to connect IMO.

File: server/command.go [hunk 2]
```diff
@@ -472,13 +472,17 @@ func (p *Plugin) executeConnectCommand(args *model.CommandArgs) (*model.CommandR
...
 		if openConnectErr != nil {
 			p.API.LogWarn("Error in checking if the user can openly connect", "user_id", args.UserId, "error", openConnectErr.Error())
 			return p.cmdError(args, genericErrorMessage)
 		}
 
 		if !canOpenlyConnect {
+			if nAvailable > 0 {
+				// spots available, but need to be on whitelist in order to connect
+				return p.cmdError(args, "You cannot connect your account at this time because an invitation is required. Please contact your system administrator to request an invitation.")
+			}
```
Comment (calebroseland) [inline server/command.go:485]: Whitelist-restrict manual connect

File: server/connect.go [hunk 3]
```diff
@@ -9,33 +9,26 @@ import (
...
 func (p *Plugin) MaybeSendInviteMessage(userID string) (bool, error) {
-	if p.getConfiguration().NewUserConnections == NewConnectionsEnabled {
-		// new connections allowed, but invites disabled
+	if p.getConfiguration().ConnectedUsersMaxPendingInvites == 0 {
 		return false, nil
 	}
 
+	if userID == p.userID {
+		return false, errors.New("cannot invite plugin bot")
+	}
```
Comment (calebroseland) [inline server/connect.go:19]: prevent plugin bot invite

File: server/connect.go [hunk 4]
```diff
@@ -9,33 +9,26 @@ import (
...
-	NewConnectionsRolloutOpen           = "rolloutOpen"
-	NewConnectionsRolloutOpenRestricted = "rolloutOpenRestricted"
-)
-
 func (p *Plugin) MaybeSendInviteMessage(userID string) (bool, error) {
-	if p.getConfiguration().NewUserConnections == NewConnectionsEnabled {
-		// new connections allowed, but invites disabled
+	if p.getConfiguration().ConnectedUsersMaxPendingInvites == 0 {
 		return false, nil
 	}
```
Comment (calebroseland) [inline server/connect.go:15]: Back to simple `max pending == 0` to disable invites

File: server/connect.go [hunk 5]
```diff
@@ -9,33 +9,26 @@ import (
...
 
-	if p.getConfiguration().NewUserConnections == NewConnectionsRolloutOpenRestricted {
-		// new connections allowed, but invites restricted to whitelist
-		isWhitelisted, whitelistErr := p.store.IsUserWhitelisted(userID)
-		if whitelistErr != nil {
-			return false, errors.Wrapf(whitelistErr, "error getting user in whitelist")
-		}
+	if user.IsBot {
+		return false, errors.Wrapf(err, "bot accounts cannot be invited")
+	}
```
Comment (calebroseland) [inline server/connect.go:28]: prevent bot-invites

File: server/connect.go [hunk 6]
```diff
@@ -9,33 +9,26 @@ import (
...
 
-		if !isWhitelisted {
-			return false, nil
-		}
+	if user.IsGuest() {
+		return false, errors.Wrapf(err, "guest accounts cannot be invited")
```
Comment (calebroseland) [inline server/connect.go:31]: prevent guest-invites

File: webapp/src/index.tsx [hunk 7]
```diff
@@ -81,6 +83,10 @@ export default class Plugin {
                 settingsEnabled = newSettingsEnabled;
                 registry.registerUserSettings?.(getSettings(serverRoute, !settingsEnabled));
             }
+
+            if (isGuest(getCurrentUserRoles(newState))) {
+                this.stopUserActivityWatch();
+            }
```
Comment (calebroseland) [inline webapp/src/index.tsx:89]: Drop notify-connect listener for guest accounts

File: server/connect.go [hunk 8]
```diff
@@ -141,25 +130,42 @@ func (p *Plugin) shouldSendInviteMessage(
...
+		isWhitelisted, err := p.store.IsUserWhitelisted(userID)
+		if err != nil {
+			return false, errors.Wrapf(err, "error in checking if user is whitelisted")
+		}
+
+		if !isWhitelisted {
+			// only whitelisted users can connect in restricted mode
+			return false, nil
+		}
+	}
```
Comment (calebroseland) [inline server/connect.go:144]: Whitelist-restrict invites

**Review by sbishel — APPROVED** [review]
We should get unit tests in where we can and probably some E2E tests as well, so we can test how everything works together.

**Comment by sbishel** [comment]
@calebroseland @lieut-data - Are we going to get this in for v1.13.0?

**Comment by lieut-data** [comment]
My apologies, @sbishel & @calebroseland: I failed to get to this review yesterday and won't be able to do it today as I'm travelling. If there's someone else who can sub in, feel free to reassign.

On the timeline: it's obviously still important but if it slips, it slips. Let's pace ourselves in all this excitement 😁

**Review by lieut-data — APPROVED** [review]

**Pull request merged**